### PR TITLE
Fix - `useIsTruncated` gives incorrect results (rect)

### LINKED ISSUE #45358: `useIsTruncated` gives incorrect results
### Describe the bug

The current [`getIsTruncated`](https://github.com/metabase/metabase/blob/54c1e07/frontend/src/metabase/hooks/use-is-truncated.ts#L41-L46) implementation uses `clientHeight`/`clientWidth`/`scrollHeight`/`scrollWidth`:
- these properties are integers and so they lose precision - it means they don't account for cases where 0px < difference < 1px
- setting `width` on an element may prevent it from resizing when ellipsis is applied, so comparing the aforementioned properties won't give any useful information

Background: [Slack thread](https://metaboat.slack.com/archives/C052LBPDAF3/p1720176721037599?thread_ts=1720096231.281389&cid=C052LBPDAF3)

### To Reproduce

https://jsfiddle.net/8b14pv96/1/

-----

1. New question > Orders
2. Add breakout "Created At" - "Month of year"
3. Click to edit breakout
4. Hover over truncated "by month of year"

There should be a tooltip.

### Information about your Metabase installation

master, 4db6ede44af77f1abce3186718400ddad232ace4

### Severity

P3

### PR DESCRIPTION
**Pull request by kamilmielnik**
Fixes #45358

### Description

The current [`getIsTruncated`](https://github.com/metabase/metabase/blob/54c1e07/frontend/src/metabase/hooks/use-is-truncated.ts#L41-L46) implementation uses `clientHeight`/`clientWidth`/`scrollHeight`/`scrollWidth`:
- these properties are integers and so they lose precision - it means they don't account for cases where 0px < difference < 1px (where "difference" means the difference between ellipsis being applied and not)
- setting `width` on an element may prevent it from resizing when ellipsis is applied, so comparing the aforementioned properties won't give any useful information

This PR changes the `getIsTruncated` to use floating-point-precise `getBoundingClientRect` instead.

It's a bit awkward to mock `getBoundingClientRect` in unit tests to test ellipsis - we have 3 cases.
I kept it for now, but in the future we should only be testing ellipsis/tooltips in e2e tests.

### How to verify

CI is green.

### CODE REVIEW

File: frontend/src/metabase/hooks/use-is-truncated.ts [hunk 1]
```diff
@@ -9,9 +9,8 @@ type UseIsTruncatedProps = {
   tolerance?: number;
 };
 
-export const useIsTruncated = <E extends Element>({
+export const useIsTruncated = <E extends HTMLElement>({
   disabled = false,
-  tolerance = 0,
```
Comment (kamilmielnik) [inline frontend/src/metabase/hooks/use-is-truncated.ts:14]: `tolerance` was never used

**Review by ranquild — CHANGES_REQUESTED** [review]
Very complex fix. Causes DOM read/write loops. There should be another way of doing this.

**Comment by kamilmielnik** [comment]
> Very complex fix. Causes DOM read/write loops. There should be another way of doing this.

I see 4 options:
1. KISS - just always show tooltips, even when text is not truncated
2. What this PR does.
3. Use Canvas to measure text (arguably more complex than what's in this PR right now - especially when `line-clamp > 1`). Something like:
    - https://github.com/Igfig/react-use-text-measurer
    - https://github.com/fabiospampinato/fast-string-truncated-width
    - https://github.com/qq15725/modern-text
4. Use programmatic text truncation instead of CSS. Something like:
    - https://github.com/pablosichert/react-truncate
    - https://github.com/ShinyChang/react-text-truncate
    - https://github.com/Irene-24/react-truncatable

@ranquild WDYT?

**Comment by ranquild** [comment]
@kamilmielnik can we add 1-2px to the original logic and accept that it would show tooltips in some edge cases?

**Comment by kamilmielnik** [comment]
> @kamilmielnik can we add 1-2px to the original logic and accept that it would show tooltips in some edge cases?

@ranquild Adding 1px would mean we'd always show tooltips, not only in edge cases. So it's [option (1)](https://github.com/metabase/metabase/pull/45347#issuecomment-2225204872). Would this be acceptable?

**Comment by kamilmielnik** [comment]
Closing in favor of #45625

**Pull request closed without merging**